File: 0101-0200/0138.Copy_List_with_Random_Pointer.java
//0ms




/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        Map<Node, Node> hashMap = new HashMap<>();
        Node cur = head;

        while (cur != null) {
            hashMap.put(cur, new Node(cur.val));
            cur = cur.next;
        }

        cur = head;

        while (cur != null) {
            Node copy = hashMap.get(cur);
            copy.next = hashMap.get(cur.next);
            copy.random = hashMap.get(cur.random);
            cur = cur.next;
        }

        return hashMap.get(head);        
    }
}




//1ms





/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        HashMap<Node, Integer> nodeToLocation = new HashMap<>();
        HashMap<Integer, Node> locationToNode = new HashMap<>();
        
        Node pointer = head;
        Node newList = null;
        int locNumber = 0;
        while(pointer!=null){
            nodeToLocation.put(pointer, locNumber);
            Node newElem = new Node(pointer.val);
            locationToNode.put(locNumber, newElem);
            if(newList == null) newList = newElem;
            else {
                locationToNode.get(locNumber-1).next = newElem;
            }
            locNumber++;
            pointer = pointer.next;
        }
        pointer = head;
        while(pointer!=null) {
            if(pointer.random==null) {
                pointer = pointer.next;
                continue;
            }
            int locThis = nodeToLocation.get(pointer);
            int locRandom = nodeToLocation.get(pointer.random);
            locationToNode.get(locThis).random = locationToNode.get(locRandom);
            pointer = pointer.next;
        }
        return newList;
    }
}

File: 0101-0200/0138.Copy_List_with_Random_Pointer.cpp
//7ms




/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        map<Node*, Node*> m;
        int i=0;
        Node* ptr = head;
        while (ptr) {
            m[ptr] =new Node(ptr->val);
            ptr = ptr->next;
        }
        ptr = head;
        while (ptr) {
            m[ptr]->next = m[ptr->next];
            m[ptr]->random = m[ptr->random];
            ptr = ptr->next;
        }
        return m[head];
    }
};





//0ms



/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/
class Solution {
public:
    Node* copyRandomList(Node* head) {
        std::unordered_map<Node*, Node*> hashMap;
        Node* cur = head;

        while (cur) {
            hashMap[cur] = new Node(cur->val);
            cur = cur->next;
        }

        cur = head;

        while (cur) {
            Node* copy = hashMap[cur];
            copy->next = hashMap[cur->next];
            copy->random = hashMap[cur->random];
            cur = cur->next;
        }

        return hashMap[head];        
    }
};




//4ms





/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(head == NULL ) return nullptr;

        unordered_map<Node*,Node*> map;
        Node* temp = head;
        Node*  head2 = new Node(-1);
        Node* prev = head2;


        while(temp){
            Node* ans = new Node(temp->val);
            map[temp]=ans;
            prev->next = ans;
            prev = ans;
            temp = temp->next;
            
        }

        temp = head;
        Node* temp2 = head2->next;
        while(temp){
        if(temp->random) temp2->random = map[temp->random];
        temp2=temp2->next;
        temp=temp->next;
            
        }

        return head2->next;
        
    }
};

